agent/containermetadata/metadata_manager.go

```diff
@@ -0,0 +1,186 @@
+// Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package containermetadata
+
+import (
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/aws/amazon-ecs-agent/agent/api"
+	"github.com/aws/amazon-ecs-agent/agent/config"
+
+	"github.com/cihub/seelog"
+	docker "github.com/fsouza/go-dockerclient"
+)
+
+const (
+	inspectContainerTimeout = 30 * time.Second
+)
+
+// MetadataManager is an interface that allows us to abstract away the metadata
+// operations
+type MetadataManager interface {
+	SetContainerInstanceArn(string)
+	CreateMetadata(*docker.HostConfig, *api.Task, *api.Container) error
+	UpdateMetadata(string, *api.Task, *api.Container) error
+	CleanTaskMetadata(*api.Task) error
+}
+
+// metadataManager implements the MetadataManager interface
+type metadataManager struct {
+	client               dockerMetadataClient
+	cluster              string
+	enabled              bool
+	dataDir              string
+	dataDirOnHost        string
+	containerInstanceARN string
+}
+
+// NewMetadataManager creates a metadataManager for a given DockerTaskEngine settings.
+func NewMetadataManager(client dockerMetadataClient, cfg *config.Config) MetadataManager {
+	return &metadataManager{
+		client:        client,
+		cluster:       cfg.Cluster,
+		enabled:       cfg.ContainerMetadataEnabled,
+		dataDir:       cfg.DataDir,
+		dataDirOnHost: cfg.DataDirOnHost,
+	}
+}
+
+// SetContainerInstanceArn sets the metadataManager's ContainerInstanceArn which is not available
+// at its creation as this information is not present immediately at the agent's startup
+func (manager *metadataManager) SetContainerInstanceArn(containerInstanceARN string) {
+	// Do nothing if disabled
+	if manager == nil || !manager.enabled {
```

Comment:
You can call methods on nil pointer receivers: https://play.golang.org/p/QMVbDa7lvo.  I think a nil check is reasonable.